conans/client/tools/settings.py

```diff
@@ -1,77 +1,382 @@
-from conans.client.build.cppstd_flags import cppstd_default
+from conans.client.build.cppstd_flags import cppstd_from_settings, cppstd_default
 from conans.errors import ConanInvalidConfiguration, ConanException
+from conans.model.version import Version
+
+
+def deduced_cppstd(conanfile):
+    """ Return either the current cppstd or the default one used by the current compiler
+
+        1. If settings.compiler.cppstd is not None - will return settings.compiler.cppstd
+        2. It settings.compiler.cppstd is None - will use compiler to deduce (reading the
+            default from cppstd_default)
+        3. If settings.compiler is None - will raise a ConanInvalidConfiguration exception
+        4. If can't detect the default cppstd for settings.compiler - will return None
+
+    :param conanfile: ConanFile instance with compiler and cppstd information
+    """
+    cppstd = cppstd_from_settings(conanfile.settings)
+    if cppstd:
+        return cppstd
+
+    compiler = conanfile.settings.get_safe("compiler")
+    if not compiler:
+        raise ConanInvalidConfiguration("Could not obtain cppstd because either compiler is not "
+                                        "specified in the profile or the 'settings' field of the "
+                                        "recipe is missing")
+    compiler_version = conanfile.settings.get_safe("compiler.version")
+    if not compiler_version:
+        raise ConanInvalidConfiguration("Could not obtain cppstd because compiler.version "
+                                        "is not specified in the profile")
+
+    return cppstd_default(conanfile.settings)
+
+
+def _remove_extension(_cppstd):
+    """ Removes gnu prefix from cppstd
+
+    :param cppstd: cppstd version
+    :return: Cppstd without extension
+    """
+    return str(_cppstd).replace("gnu", "")
+
+
+def _contains_extension(cppstd):
+    """ Returns if cppstd contains gnu extension
+
+    :param cppstd: cppstd version
+    :return: True, if current cppstd contains the gnu extension. Otherwise, False.
+    """
+    return str(cppstd).startswith("gnu") and _remove_extension(cppstd).isdigit()
```

Comment:
why is digit? what about 0x, 2a, 1z and so on?

Reply:
`cppstd` never contain 0x, 2a, 1z. `cppstd` - the value in settings

Reply:
nice!